### ISSUE TITLE
toObject not removing functions

### ISSUE THREAD

**Issue by gkaimakas** [balderdashy/sails#2610]
Hello,

I have the following model:

```
var crypto = require("crypto");
var Promise = require("bluebird");
var moment = require("moment");
var accountConf = require("./config/AccountConf");
var globalConf = sails.config.globals;


/**
 * @module models/account
 * @author George Kaimakas
 * @description A model representing an account with different roles and statuses
 */

/**
 * @typedef accountJSON
 * @type {object}
 * @property {string} id the account id
 * @property {string} firstname
 * @property {string} lastname
 * @property {string} username the email the user used to create the account
 * @property {string} status valid values are pending, active, inactive, revoked
 * @property {array} roles can have many roles, some are user, avertiser, admin
 * @property {date} createdAt when it was created
 * @property {date} updatedAt last updated at
 * @property {date} verifiedAt verified at
 */

module.exports = {
  tableName     : "account",
  autoPK        : true,
  autoCreatedAt : true,
  autoUpdatedAt : true,
  schema        : true,

  /**
   * @memberof module:models/account#
   * @alias Account
   * @property {string} firstname
   * @property {string} lastname
   * @property {string} username a valid email
   * @property {string} password it is encrypted before saving.
   * @property {int} status
   * @property {int} roles
   * @property {array} transactions the transaction associated with that account
   * @property {date} verifiedAt
   */
  attributes : {
    firstname : {
      type     : "string",
      required : false
    },

    lastname : {
      type : "string",
      required : false
    },

    username  : {
      type     : "string",
      email    : true,
      required : true,
      notNull  : true,
      unique   : true
    },


    password : {
      type      : "string",
      required  : true,
      notNull   : true,
      minLength : globalConf.passwordLengthMin,
      maxLength : globalConf.passwordLengthMax
    },

    //passwordConfirm : {
    //  type : "string"
    //}

    /**
     * account status
     */
    status : {
      type       : "int",
      required   : true,
      defaultsTo : accountConf.status.pending
    },

    /**
     * account roles. An account can have many roles. Roles are integers defined in sails/config/roles
     */
    roles : {
      type       : "int",
      required   : true,
      defaultsTo : accountConf.roles.user
    },

    /**
     * the transaction associated with this account
     */
    transactions : {
      collection : "transaction",
      via        : "account"
    },

    verifiedAt : {
      type : "date"
    },

    /**
     * @memberof module:models/account#
     * @method isUser
     * @description checks if the account has the user role set.
     * @returns {boolean} true if the account has the user role, false otherwise.
     */
    isUser : function () {
      return this.hasRole(accountConf.roles.user);
    },

    /**
     * @memberof module:models/account#
     * @func isAdvertiser
     * @desc checks if the account has the advertiser role set
     * @returns {boolean}
     */
    isAdvertiser : function () {
      return this.hasRole(accountConf.roles.advertiser);
    },

    /**
     * @memberof module:models/account#
     * @func isAdmin
     * @description checks if the account has the admin role ser
     * @returns {boolean}
     */
    isAdmin : function () {
      return this.hasRole(accountConf.roles.admin);
    },

    /**
     * @memberof module:models/account#
     * @func setAdvertiser
     * @desc adds advertiser role to the account
     * @returns {account}
     */
    setAdvertiser : function () {
      return this.addRole(accountConf.roles.advertiser);
    },

    /**
     * @memberof module:models/account#
     * @funct setAdmin
     * @desc adds admin role to the account
     * @returns {account}
     */
    setAdmin : function () {
      return this.addRole(accountConf.roles.admin);
    },

    /**
     * removes the advertiser role from an account instance
     * @memberof module:models/account#
     * @func removeAdvertiser
     * @returns {account}
     */
    removeAdvertiser : function () {
      return this.removeRole(accountConf.roles.advertiser);
    },

    /**
     * removes the admin role
     * @memberof module:models/account#
     * @func removeAdmin
     * @returns {account}
     */
    removeAdmin : function () {
      return this.removeRole(accountConf.roles.admin);
    },


    /**
     * changes the status to active
     * @memberof module:models/account#
     * @returns {account}
     */
    setStatusActive : function () {
      return this.setStatus(accountConf.status.active);
    },

    /**
     * changes the stauts to inactive
     * @memberof module:models/account#
     * @returns {account}
     */
    setStatusInactive : function () {
      return this.setStatus(accountConf.status.inactive);
    },

    /**
     * changes the status to revoked
     * @memberof module:models/account#
     * @returns {account}
     */
    setStatusRevoked : function () {
      return this.setStatus(accountConf.status.revoked);
    },

    /**
     * checks if the account has the specified role
     * @memberof module:models/account#
     * @param {int} role
     * @returns {boolean}
     */
    hasRole : function (role) {
      return (this.roles & role) === role;
    },

    /**
     * adds a new role to the account
     * @memberof module:models/account#
     * @param role
     * @return {account}
     */
    addRole : function (role) {
      this.roles = (this.roles | role);
      return this;
    },

    /**
     * removes a role from the account
     * @memberof module:models/account#
     * @param role
     * @returns {object}
     */
    removeRole : function (role) {
      this.roles = (this.roles & (~role));
      return this;
    },

    /**
     * checks if the status is pending
     * @memberof module:models/account#
     * @returns {boolean}
     */
    isPending : function () {
      return this.status == accountConf.status.pending;
    },

    /**
     * checks if the status is active
     * @memberof module:models/account#
     * @returns {boolean}
     */
    isActive : function () {
      return this.status === accountConf.status.active;
    },

    /**
     * checks if the status is inactive
     * @memberof module:models/account#
     * @returns {boolean}
     */
    isInactive : function () {
      return this.status === accountConf.status.inactive;
    },

    /**
     * checks if the status is revoked
     * @memberof module:models/account#
     * @returns {boolean}
     */
    isRevoked : function () {
      return this.status === accountConf.status.revoked
    },


    /**
     * sets the status
     * @memberof module:models/account#
     * @returns {account}
     */
    setStatus : function (status) {
      this.status = status;
      return this;
    },

    /**
     * returns the status of the object as a string in lower case
     * @returns {string}
     */
    statusToString : function () {
      return Strings.attrToString(accountConf.status, this.status);
    },

    /**
     * verifies that the password matches the saved password
     * @param {string} password the password to check if matches
     * @returns {boolean}
     */
    verifyPassword : function (password) {
      var cipher = crypto.createCipher(
        sails.config.cryptoAlgorithm || process.env.CRYPTO_ALGORITHM,
        sails.config.cryptoKey || process.env.CRYPTO_KEY
      );

      var _password = cipher.update(password, "utf8", "hex") + cipher.final("hex");
      return _password == this.password;
    },

    /**
     * changes the account password
     * @param oldPassword
     * @param newPassword
     * @returns {Promise}
     */
    changePassword : function (oldPassword, newPassword) {
      if (this.verifyPassword(oldPassword)) {
        this.password = newPassword;
        this.passwordConfirm = newPassword;
        return this.save();
      } else {
        var err = new Error(Strings.password_change_match_err);
        err.status = sails.config.responses.conflict;
        return Promise.reject(err);
      }
    },

    /**
     * turns an account with status pending to an account with status active while setting
     * the verifiedAt field at the current date
     * @returns {Promise}
     */
    verifyAccount : function () {
      if (this.isPending()) {
        this.verifiedAt = moment().format();
        this.setStatusActive();
        return this.save();
      } else {
        var err = new Error(Strings.account_verify_not_pending_err);
        err.status = sails.config.responses.conflict;
        return Promise.reject(err);
      }
    },

    /**
     * Converts the account entity to a valid JSON object
     * @memberof module:models/account#
     * @returns {accountJSON} account the account entity as a json object
     */
    toJSON : function () {
      var obj = this.toObject();
      delete obj.password;

      obj.roles = [];
      this.isUser() ? obj.roles.push("user") : 0;
      this.isAdvertiser() ? obj.roles.push("advertiser") : 0;
      this.isAdmin() ? obj.roles.push("admin") : 0;

      obj.status = this.statusToString();

      return obj;
    }
  },

  /**
   * @description Changes the password of an account
   * @param accountId {string} the id of the account
   * @param oldPassword {string} the current password, unencrypted
   * @param newPassword {string} the new password, unencrypted
   * @returns {Promise} a promise to change the password
   */
  changePassword : function (accountId, oldPassword, newPassword) {

    return this.findOne({
      id : accountId
    })
      .then(function (account) {
        return account.changePassword(oldPassword, newPassword);
      })
      .catch(function (err) {
        return Promise.reject(err);
      });
  },

  beforeValidate : function (account, next) {

    /* istanbul ignore next */
    if (!account.passwordConfirm) {
      next(Strings.password_confirm_null_err);
    }

    /* istanbul ignore next */
    if (account.password != account.passwordConfirm) {
      next(Strings.password_confirm_match_err);
    }

    next();
  },


  beforeCreate : function (account, next) {
    /* istanbul ignore next */
    var cipher = crypto.createCipher(
      sails.config.cryptoAlgorithm || process.env.CRYPTO_ALGORITHM,
      sails.config.cryptoKey || process.env.CRYPTO_KEY
    );
    account.password = cipher.update(account.password, "utf8", "hex") + cipher.final("hex");
    next();
  },

  beforeUpdate : function (account, next) {
    var cipher = crypto.createCipher(
      sails.config.cryptoAlgorithm || process.env.CRYPTO_ALGORITHM,
      sails.config.cryptoKey || process.env.CRYPTO_KEY
    );
    account.password = cipher.update(account.password, "utf8", "hex") + cipher.final("hex");
    next();
  }
};
```

and whenever i call toJSON or toObject I still get the functions...

Is this a bug ?

**Comment by sgress454** [comment]
Thanks @gkaimakas, this does look to be a bug.  We're looking into it!

**Comment by gkaimakas** [comment]
Cool!

**Comment by sailsbot** [comment]
Thanks for posting, @gkaimakas.  I'm a repo bot-- nice to meet you!

It has been 60 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by sgress454** [comment]
FYI should be fixed by https://github.com/balderdashy/waterline/pull/969
